### ISSUE TITLE
DTW crashes in Max 8

### ISSUE THREAD

**Issue by Knallberto** [irllabs/ml-lib#177]
Hi, I tried usind the dynamic time warping example you provide, but when I want to start 'mapping' Max crashes. I don't know why this happens, I get no error message. I am working with Max 8 and "ml.dtw 1.1.2: Dynamic Time Warping based on the GRT library version 0.2.5 revision: 2-Jan-2017".

**Comment by NiccoloGranieri** [comment]
Hi @Knallberto, I just tested [ml.dtw] on Max 8 both from scratch and using the provided example and I am unable to reproduce your issue. Could you provide us with a bit more info on your setup? Are you using the [ml.lib] version found on the Max Package Manager? Are you running Big Sur? What is the exact procedure you need to follow to make the app crash? Thanks in advance for your time!

**Comment by Knallberto** [comment]
Hi, thanks for the reply and thanks for your time! 
I am running Windows 10 Home with a i5-4440 CPU @ 3.1GHz, 16GB Ram. I installed ml.lib with the max packages manager (ml.lib by IRL Labs 1.1.2) in Max 8 (64-bit). 
First I tried with Data from a MMA8452 Sensor with Arduino and serial communication. I pressed the clear button, chose train in the dropdown menu, selecteded the first button, clicked listen, made a gesture with the sensor (circle) and then stopped the listen button. I repeated the same procedure for two more slots, then clicked train and selected map in the drop-down menu. I clicked on listen and after a few seconds max crashed. I tried this a few times and everytime I got in the map mode and started listening it crashed. Next, I tried exactly the same procedure with data from the accellerometer in my iphone using GyrOSC and udpreceive (like in the example). This also crahes after a few seconds of incoming Data in the mapping mode. I did not change any of the default settings.

**Comment by NiccoloGranieri** [comment]
Hi @Knallberto, I just tested this on a Windows 10 machine and unfortunately (or luckily?) I can't reproduce the crash. Before we go down the log-rabbit hole could you please do me a favour and uninstall ml.lib, restart your laptop, and re-install a fresh copy? If the good old "have you tried turning it off and on again" doesn't work, I'll probably ask you to send me some logs.

**Comment by Knallberto** [comment]
I installes a fresh copy and restarted my computer, still the same problem. When I stop the 'listen' in mapping mode fast enough, before max crashes, I get the message: "error: ml.dtw: unable to map input". How do I create logs?

**Comment by NiccoloGranieri** [comment]
Ok, so make it crash, and find the crash log here:

`C:\Users\yourusername\AppData\Roaming\Cycling '74\Logs`

Send me the latest one and I'll see what I can do.

**Comment by Knallberto** [comment]
I tried to make it crash to create a logfile, but no log file is created. Strangely, there is a logfile from yester in the folder that probably includes the same information (?). So the crash goes like this: first the window 'freezes', when I try to do something it gets greyed out and a message from windows appears "program is not responding. When I close the program and restart I get the "Looks like Max previously crashed. Would you like to recover your work?" message. But no logfile. I attatched the logfile from yesterday, maybe there is some useful information. 
[Max-2021-02-15_22-52-43-667771f.zip](https://github.com/irllabs/ml-lib/files/5989149/Max-2021-02-15_22-52-43-667771f.zip)

**Comment by NiccoloGranieri** [comment]
So after having a chat with the lead dev on this project it seems like Max is hanging not crashing. Could you please tell us:

- Does this happen only with [ml.dtw] or also with other objects? (You could try [ml.ann] for example)
- Could you check what the CPU load is when Max hangs?

**Comment by Knallberto** [comment]
With ml.ann I am not sure if I am doing it right. I gave examples for different presets (so first I selected a preset, then I hit the toggle, then turned off the toggle and selected another preset and so forth) for the soundz patcher, is that correct? I receive a message "error: invalid input length, expected 8, got 3" when I select map. I also tried the "4-Pose_Recognition-SVM.maxpat" - here the class output seems to be totally wrong, but I do not get an error message and there is no crash or 'hanging'. 

CPU load is around 30% 
<img width="255" alt="cpu_load_max" src="https://user-images.githubusercontent.com/34421732/108086001-36a27380-7076-11eb-9e5f-c565723877ae.PNG">

**Comment by NiccoloGranieri** [comment]
Yeah I just saw that the [ml.ann] example is missing a few steps. If you want to try it, I'll attach an updated version but I don't think you're going to have issues with it.
[6-Multi-Dimenional_Mapping-ANN.maxpat.zip](https://github.com/irllabs/ml-lib/files/5989966/6-Multi-Dimenional_Mapping-ANN.maxpat.zip)

**Comment by Knallberto** [comment]
Thanks for the patch, this is working perfectly

**Comment by Knallberto** [comment]
I tried to use the DTW patch on another computer and there I get the same problem.

**Comment by NiccoloGranieri** [comment]
That's interesting @Knallberto. Before we dive deep into DTW, would you mind saving the ml.dtw data and sending it over together with a few seconds worth of data streaming from your phone? You could record the data into a test file using the [coll] object. To save the ML file instead, just pass a (write) message to the ml object specifying the path.

**Comment by Knallberto** [comment]
Thanks for the reply, I tried to record the data, but I do not know how to do it. When I specify the path with the "write" message, how exactly does it have to look? when I just type something like c:/folder/filename I get a "unable to write model to path" message. And I also do not know how the coll object works, do I have to prepend a number (1,2,3 ...) to ever list I want to record?

**Comment by NiccoloGranieri** [comment]
Here you go @Knallberto, use this to generate a ".data" file and a ".txt" file, and send those back.
[dataRecording.maxpat.zip](https://github.com/irllabs/ml-lib/files/5997353/dataRecording.maxpat.zip)

**Comment by Knallberto** [comment]
Thanks for the patch, here is the data:
[DTW_data.zip](https://github.com/irllabs/ml-lib/files/5997847/DTW_data.zip)

**Comment by NiccoloGranieri** [comment]
Hi @Knallberto, having a look at the data you sent over nothing seems to be wrong. However, I managed to make Max _hang_ by mapping continuous data for a bunch of seconds. So my question is: when you map data, are you delimitating the "gestures" by turning on and off the listening toggle, or are you turning the listening on expecting DTW to recognise and segment the individual gestures previously trained?

**Comment by Knallberto** [comment]
Hi, I see, I am "turning the listening on expecting DTW to recognise and segment the individual gestures previously trained" - is this not possible?

**Comment by NiccoloGranieri** [comment]
No, unfortunately not. However, you could implement something to make it work that way in your patch. For example, let's say you're trying to recognise shapes drawn on a drawing tablet. You could use the act of touching the tablet with the pen as "turn listening on", and turn it off when the pen is lifted. This could be anything, it really depends on the devices/hardware you're using.

**Comment by Knallberto** [comment]
Okay, sorry I did not know this. Thank you for your time! I was thinking of using a IMU Sensor with x/y/z information to recognise gestures while moving it.

**Comment by Knallberto** [comment]
I have one more question on the Pose-Recognition-SVM. For testing I used three distinct messages (1 2 3, 1 1 1, 1 5 4). I clicked 'clear', then selected 'train', then I sent in the first message and selected 'class 1' - the output from 'p train-predict' is 'add 1 1 2 3'. I did the same for the next two messages and class 2 and three. Lastly, I klicked train and selected the map mode. Now, when I send in the first message, the output from ml.svm is 3, when I send in the second message the output is three and when I send in the third message the output is two. I anticipated the outcome first message = class one, second message = class two and third message = class 3. Am I doing something wrong here, too ?
